# 1.Getting & Knowing Data

## Q1. 데이터를 로드하라. 데이터는 \t을 기준으로 구분

In [1]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/pandas/main/lol.csv', sep = '\t')
df.head()

,gameId,creationTime,gameDuration,seasonId,winner,firstBlood,firstTower,firstInhibitor,firstBaron,firstDragon,...,t2_towerKills,t2_inhibitorKills,t2_baronKills,t2_dragonKills,t2_riftHeraldKills,t2_ban1,t2_ban2,t2_ban3,t2_ban4,t2_ban5
0,3326086514,1504279457970,1949,9,1,2,1,1,1,1,...,5,0,0,1,1,114,67,43,16,51
1,3229566029,1497848803862,1851,9,1,1,1,1,0,1,...,2,0,0,0,0,11,67,238,51,420
2,3327363504,1504360103310,1493,9,1,2,1,1,1,2,...,2,0,0,1,0,157,238,121,57,28
3,3326856598,1504348503996,1758,9,1,1,1,1,1,1,...,0,0,0,0,0,164,18,141,40,51
4,3330080762,1504554410899,2094,9,1,2,1,1,1,1,...,3,0,0,1,0,86,11,201,122,18


## Q2.데이터의 상위 5개 행을 출력하라

## Q3.데이터의 행과 열의 갯수를 파악하라

## Q4.전체 컬럼을 출력하라

## Q5.6번째 컬럼명을 출력하라

## Q6.6번째 컬럼의 데이터 타입을 확인하라

## Q7.데이터셋의 인덱스 구성은 어떤가

## Q8.6번째 컬럼의 3번째 값은 무엇인가

2.제주 날씨, 인구에 따른 교통량 데이터

## Q9.데이터를 로드하라. 컬럼이 한글이기에 적절한 처리

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/pandas/main/Jeju.csv', encoding = 'euc-kr')
df.head(5)

,id,일자,시도명,읍면동명,거주인구,근무인구,방문인구,총 유동인구,평균 속도,평균 소요 시간,평균 기온,일강수량,평균 풍속
0,22448,2018-01-01,제주시,도두동,32249.987,3418.266,102709.092,138377.345,39.556,29.167,5.0,0.0,2.5
1,22449,2018-01-01,제주시,외도동,213500.997,10341.172,112692.789,336534.958,32.900,30.900,5.0,0.0,2.5
2,22450,2018-01-01,제주시,이도2동,1212382.218,96920.834,541194.481,1850497.533,29.538,35.692,2.9,0.0,2.4
3,22451,2018-01-01,제주시,일도1동,33991.653,6034.253,72155.919,112181.825,30.000,23.500,2.9,0.0,2.4
4,22452,2018-01-01,서귀포시,대천동,155036.925,9403.969,150882.409,315323.303,41.583,14.375,5.1,0.0,2.3


## Q10.데이터 마지막 3개행을 출력하라

## Q11.수치형 변수를 가진 컬럼을 출력하라

## Q12.범주형 변수를 가진 컬럼을 출력하라

## Q13.각 컬럼의 결측치 숫자를 파악하라

## Q14.각 컬럼의 데이터 수, 데이터타입을 한 번에 확인하라

## Q15.각 수치형 변수의 분포(사분위, 평균, 표준편차, 최대, 최소)를 확인하라

## Q16.거주 인구 컬럼의 값들을 출력하라

## Q17.평균 속도 컬럼의 4분위 범위(IQR) 값을 구하여라

## Q18.읍면동명 컬럼의 유일값 갯수를 출력하라

## Q19.읍면동명 컬럼의 유일값을 모두 출력하라

# 2.Filtering & Sorting

## Q20.데이터를 로드하라

In [3]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/pandas/main/chipo.csv')
df.head()

,order_id,quantity,item_name,choice_description,item_price
0,1,1,Chips and Fresh Tomato Salsa,NaN,$2.39
1,1,1,Izze,[Clementine],$3.39
2,1,1,Nantucket Nectar,[Apple],$3.39
3,1,1,Chips and Tomatillo-Green Chili Salsa,NaN,$2.39
4,2,2,Chicken Bowl,"[Tomatillo-Red Chili Salsa (Hot), [Black Beans...",$16.98


## Q21.quantity컬럼 값이 3인 데이터를 추출하여 첫 5행을 출력하라

## Q22.quantity컬럼 값이 3인 데이터를 추출하여 index를 0부터 정렬하고 첫 5행을 출력하라

## Q23.quantity, item_price 두 개의 컬럼으로 구성된 새로운 데이터 프레임을 정의하라

## Q24.item_price 컬럼의 달러표시 문자를 제거하고, float 타입으로 저장하여 new_price 컬럼에 저장하라

## Q25.new_price 컬럼이 5이하의 값을 가지는 데이터프레임을 추출하고, 전체 갯수를 구하여라

## Q26.item_name명이 Chicken Salad Bowl인 데이터 프레임을 추출하고 index 값을 초기화 하여라

## Q27.new_price 값이 9이하이고 item_name 값이 Chicken Salad Bowl인 데이터 프레임을 추출하라

## Q28.df의 new_price 컬럼 값에 따라 오름차순으로 정리하고 index를 초기화 하여라

## Q29.df의 item_name컬럼 값 중 Chips 포함하는 경우의 데이터를 출력하라

## Q30.df의 짝수번째 컬럼만을 포함하는 데이터프레임을 출력하

## Q31.df의 new_price 컬럼 값에 따라 내림차순으로 정리하고 index를 초기화 하여라

## Q32.df의 item_name컬럼 값이 Steak Salad 또는 Bowl인 데이터를 인덱싱하라

## Q33.df의 item_name 컬럼 값이 Steak Salad 또는 Bowl인 데이터를 데이터 프레임화 한 후, item_name를 기준으로 중복행이 있으면 제거하되 첫번째 케이스만 남겨라

## Q34.df의 item_name 컬럼 값이 Steak Salad 또는 Bowl인 데이터를 데이터 프레임화 한 후, item_name를 기준으로 중복행이 있으면 제거하되 마지막 케이스만 남겨라

## Q35.df의 데이터 중 new_price값이 new_price값의 평균 값 이상을 가지는 데이터들을 인덱싱하라

## Q36.df의 데이터 중 item_name의 값이 Izze인 데이터를 Fizzy Lizzy로 수정하라

## Q37.df의 데이터 중 choice_description값이 NaN인 데이터의 갯수를 구하여라

## Q38.df의 데이터 중 choice_description값이 NaN인 데이터를 NoData로 대체하라(loc이용)

## Q39.df의 데이터 중 choice_description값에 black이 들어가는 경우를 인덱싱하라

## Q40.df의 데이터 중 choice_description값에 Vegetables들어가지 않는 경우의 갯수를 출력

## Q41.df의 데이터 중 item_name값이 N으로 시작하는 데이터를 모두 출력하라

## Q42.df의 데이터 중 item_name 값의 단어 갯수가 15개 이상인 데이터를 인덱싱하라

## Q43.df의 데이터 중 new_price값이 lst에 해당하는 경우의 데이터 프레임을 구하고 그 갯수를 출력하라

lst =[1.69, 2.39, 3.39, 4.45, 9.25, 10.98, 11.75, 16.98]

# 3.Grouping

## Q44.데이터를 로드하고 상위 5개 컬럼을 출력하라

In [4]:
import pandas as pd
df= pd.read_csv('https://raw.githubusercontent.com/Datamanim/pandas/main/AB_NYC_2019.csv')
df.head()

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
0,2539,Clean & quiet apt home by the park,2787,John,Brooklyn,Kensington,40.64749,-73.97237,Private room,149,1,9,2018-10-19,0.21,6,365
1,2595,Skylit Midtown Castle,2845,Jennifer,Manhattan,Midtown,40.75362,-73.98377,Entire home/apt,225,1,45,2019-05-21,0.38,2,355
2,3647,THE VILLAGE OF HARLEM....NEW YORK !,4632,Elisabeth,Manhattan,Harlem,40.80902,-73.94190,Private room,150,3,0,NaN,NaN,1,365
3,3831,Cozy Entire Floor of Brownstone,4869,LisaRoxanne,Brooklyn,Clinton Hill,40.68514,-73.95976,Entire home/apt,89,1,270,2019-07-05,4.64,1,194
4,5022,Entire Apt: Spacious Studio/Loft by central park,7192,Laura,Manhattan,East Harlem,40.79851,-73.94399,Entire home/apt,80,10,9,2018-11-19,0.10,1,0


## Q45.데이터의 각 host_name의 빈도수를 구하고 host_name으로 정렬하여 상위 5개를 출력하라

## Q46.데이터의 각 host_name의 빈도수를 구하고 빈도수 기준 내림차순 정렬한 데이터 프레임을 만들어라. 빈도수 컬럼은 counts로 하여라

## Q47.neighbourhood_group의 값에 따른 neighbourhood 컬럼 값의 갯수를 구하여라

## Q48.neighbourhood_group의 값에 따른 neighbourhood컬럼 값 중 neighbourhood_group그룹의 최댓값들을 출력하라

## Q49.neighbourhood_group 값에 따른 price값의 평균, 분산, 최대, 최소 값을 구하여라

## Q50.neighbourhood_group 값에 따른 reviews_per_month 평균, 분산, 최대, 최소 값을 구하여라

## Q51.neighbourhood값과 neighbourhood_group 값에 따른 price의 평균을 구하여라

## Q52.neighbourhood 값과 neighbourhood_group 값에 따른 price 의 평균을 계층적 indexing 없이 구하라

## Q53.neighbourhood 값과 neighbourhood_group 값에 따른 price 의 평균을 계층적 indexing 없이 구하고 nan 값은 -999값으로 채워라

## Q54.데이터중 neighbourhood_group 값이 Queens값을 가지는 데이터들 중 neighbourhood 그룹별로 price값의 평균, 분산, 최대, 최소값을 구하라

## Q55.데이터중 neighbourhood_group 값에 따른 room_type 컬럼의 숫자를 구하고 neighbourhood_group 값을 기준으로 각 값의 비율을 구하여라

# 4.Apply & Map

## Q56.데이터를 로드하고 데이터 행과 열의 갯수를 출력하라

In [5]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/pandas/main/BankChurnersUp.csv', index_col = 0)
df.head()

,CLIENTNUM,Attrition_Flag,Customer_Age,Gender,Dependent_count,Education_Level,Marital_Status,Income_Category,Card_Category,Months_on_book,Total_Relationship_Count,Months_Inactive_12_mon,Contacts_Count_12_mon,Credit_Limit,Total_Revolving_Bal,Avg_Open_To_Buy,Total_Amt_Chng_Q4_Q1,Total_Trans_Amt
0,768805383,Existing Customer,45,M,3,High School,Married,$60K - $80K,Blue,39,5,1,3,12691.0,777,11914.0,1.335,1144
1,818770008,Existing Customer,49,F,5,Graduate,Single,Less than $40K,Blue,44,6,1,2,8256.0,864,7392.0,1.541,1291
2,713982108,Existing Customer,51,M,3,Graduate,Married,$80K - $120K,Blue,36,4,1,0,3418.0,0,3418.0,2.594,1887
3,769911858,Existing Customer,40,F,4,High School,Unknown,Less than $40K,Blue,34,3,4,1,3313.0,2517,796.0,1.405,1171
4,709106358,Existing Customer,40,M,3,Uneducated,Married,$60K - $80K,Blue,21,5,1,0,4716.0,0,4716.0,2.175,816


## Q57.Income_Category의 카테고리를 map 함수를 이용하여 다음과 같이 변경하여 newIncome 컬럼에 매핑하라

Unknown : N
Less than $40K : a
$40K - $60K : b
$60K - $80K : c
$80K - $120K : d
$120K +’ : e

## Q58.Income_Category의 카테고리를 apply 함수를 이용하여 다음과 같이 변경하여 newIncome 컬럼에 매핑하라
Unknown : N
Less than $40K : a
$40K - $60K : b
$60K - $80K : c
$80K - $120K : d
$120K +’ : e

## Q59.Customer_Age의 값을 이용하여 나이 구간을 AgeState 컬럼으로 정의하라. (0~9 : 0 , 10 ~ 19 :10 , 20~29 :20 …) 각 구간의 빈도수를 출력하라

## Q60.Education_Level의 값중 Graduate단어가 포함되는 값은 1 그렇지 않은 경우에는 0으로 변경하여 newEduLevel 컬럼을 정의하고 빈도수를 출력하라

## Q61.Credit_Limit 컬럼값이 4500 이상인 경우 1 그외의 경우에는 모두 0으로 하는 newLimit 정의하라. newLimit 각 값들의 빈도수를 출력하라

## Q62.Marital_Status 컬럼값이 Married 이고 Card_Category 컬럼의 값이 Platinum인 경우 1 그외의 경우에는 모두 0으로 하는 newState컬럼을 정의하라. newState의 각 값들의 빈도수를 출력하라

## Q63.Gender 컬럼값 M인 경우 male F인 경우 female로 값을 변경하여 Gender 컬럼에 새롭게 정의하라. 각 value의 빈도를 출력하라

# 5.Time Series

## Q64.데이터를 로드하고 각 열의 데이터 타입을 파악하라

In [6]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/pandas/main/timeTest.csv')
df.head()

,Yr_Mo_Dy,RPT,VAL,ROS,KIL,SHA,BIR,DUB,CLA,MUL,CLO,BEL,MAL
0,2061-01-01,15.04,14.96,13.17,9.29,NaN,9.87,13.67,10.25,10.83,12.58,18.50,15.04
1,2061-01-02,14.71,NaN,10.83,6.50,12.62,7.67,11.50,10.04,9.79,9.67,17.54,13.83
2,2061-01-03,18.50,16.88,12.33,10.13,11.17,6.17,11.25,NaN,8.50,7.67,12.75,12.71
3,2061-01-04,10.58,6.63,11.75,4.58,4.54,2.88,8.63,1.79,5.83,5.88,5.46,10.88
4,2061-01-05,13.33,13.25,11.42,6.17,10.71,8.21,11.92,6.54,10.92,10.34,12.92,11.83


## Q65.Yr_Mo_Dy을 판다스에서 인식할 수 있는 datetime64타입으로 변경하라

## Q66.Yr_Mo_Dy에 존재하는 년도의 유일값을 모두 출력하라

## Q67.Yr_Mo_Dy에 년도가 2061년 이상의 경우에는 모두 잘못된 데이터이다. 해당경우의 값은 100을 빼서 새롭게 날짜를 Yr_Mo_Dy 컬럼에 정의하라

## Q68.년도별 각컬럼의 평균값을 구하여라

## Q69.weekday컬럼을 만들고 요일별로 매핑하라 ( 월요일: 0 ~ 일요일 :6)

## Q70.weekday컬럼을 기준으로 주말이면 1 평일이면 0의 값을 가지는 WeekCheck 컬럼을 만들어라

## Q71.년도, 일자 상관없이 모든 컬럼의 각 달의 평균을 구하여라

## Q72.모든 결측치는 컬럼기준 직전의 값으로 대체하고 첫번째 행에 결측치가 있을경우 뒤에있는 값으로 대채하라

## Q73.년도 - 월을 기준으로 모든 컬럼의 평균값을 구하여라

## Q74.RPT 컬럼의 값을 일자별 기준으로 1차차분하라

## Q75.RPT와 VAL의 컬럼을 일주일 간격으로 각각 이동평균한값을 구하여라

서울시 미세먼지 데이터

In [7]:
df =pd.read_csv('https://raw.githubusercontent.com/Datamanim/pandas/main/seoul_pm.csv')
df.head()

,(년-월-일:시),PM10등급,PM10,PM2.5등급,PM2.5,오존등급,오존,이산화질소등급,이산화질소,일산화탄소등급,일산화탄소,아황산가스등급,아황산가스
0,2021-05-15:15,보통,47.0,보통,19.0,좋음,0.017,좋음,0.023,좋음,0.4,좋음,0.003
1,2021-05-15:14,보통,43.0,보통,20.0,좋음,0.024,좋음,0.019,좋음,0.3,좋음,0.003
2,2021-05-15:13,보통,34.0,보통,24.0,보통,0.035,좋음,0.017,좋음,0.4,좋음,0.004
3,2021-05-15:12,보통,41.0,보통,27.0,보통,0.037,좋음,0.020,좋음,0.4,좋음,0.004
4,2021-05-15:11,보통,51.0,보통,34.0,보통,0.033,좋음,0.023,좋음,0.4,좋음,0.005


## Q76.년-월-일:시 컬럼을 pandas에서 인식할 수 있는 datetime 형태로 변경하라. 서울시의 제공데이터의 경우 0시가 24시로 표현된다

## Q77.일자별 영어요일 이름을 dayName 컬럼에 저장하라

## Q78.일자별 각 PM10등급의 빈도수를 파악하라

## Q79.시간이 연속적으로 존재하며 결측치가 없는지 확인하라

시간을 차분했을 경우 첫 값은 nan, 이후 모든 차분값이 동일하면 연속이라 판단한다.

## Q80.오전 10시와 오후 10시(22시)의 PM10의 평균값을 각각 구하여라

## Q81.날짜 컬럼을 index로 만들어라

## Q82.데이터를 주단위로 뽑아서 최소,최대 평균, 표준표차를 구하여라

# 6.pivot

In [10]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/pandas/main/under5MortalityRate.csv')
df.head()

,Location,Period,Indicator,Dim1,First Tooltip
0,Afghanistan,2019,Under-five mortality rate (probability of dyin...,Both sexes,60.27 [47.44-74.62]
1,Afghanistan,2019,Under-five mortality rate (probability of dyin...,Male,63.83 [50.04-79.16]
2,Afghanistan,2019,Under-five mortality rate (probability of dyin...,Female,56.57 [44.54-70.18]
3,Afghanistan,2018,Under-five mortality rate (probability of dyin...,Both sexes,62.54 [50.64-75.41]
4,Afghanistan,2018,Under-five mortality rate (probability of dyin...,Male,66.08 [53.41-79.76]


## Q83.Indicator을 삭제하고 First Tooltip 컬럼에서 신뢰구간에 해당하는 표현을 지워라



## Q84.년도가 2015년 이상, Dim1이 Both sexes인 케이스만 추출하라

## Q85. 문제에서 추출한 데이터로 아래와 같이 나라에 따른 년도별 사망률을 데이터 프레임화 하라

## Q86.Dim1에 따른 년도별 사망비율의 평균을 구하라

올림픽 메달리스트 정보 데이터

In [9]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/pandas/main/winter.csv')
df.head()

,Year,City,Sport,Discipline,Athlete,Country,Gender,Event,Medal
0,1924,Chamonix,Biathlon,Biathlon,"BERTHET, G.",FRA,Men,Military Patrol,Bronze
1,1924,Chamonix,Biathlon,Biathlon,"MANDRILLON, C.",FRA,Men,Military Patrol,Bronze
2,1924,Chamonix,Biathlon,Biathlon,"MANDRILLON, Maurice",FRA,Men,Military Patrol,Bronze
3,1924,Chamonix,Biathlon,Biathlon,"VANDELLE, André",FRA,Men,Military Patrol,Bronze
4,1924,Chamonix,Biathlon,Biathlon,"AUFDENBLATTEN, Adolf",SUI,Men,Military Patrol,Gold


## Q87.데이터에서 한국 KOR 데이터만 추출하라

## Q88.한국 올림픽 메달리스트 데이터에서 년도에 따른 medal 갯수를 데이터프레임화 하라

## Q89.전체 데이터에서 sport종류에 따른 성별수를 구하여라

## Q90.전체 데이터에서 Discipline종류에 따른 따른 Medal수를 구하여라